src/main/java/net/haesleinhuepf/clijx/plugins/DeconvolveRichardsonLucyFFT.java

```diff
@@ -0,0 +1,215 @@
+import net.haesleinhuepf.clij2.CLIJ2;
+import net.haesleinhuepf.clij2.utilities.HasAuthor;
+import net.imagej.ops.OpService;
+import net.imagej.ops.filter.pad.DefaultPadInputFFT;
+import net.imagej.ops.filter.pad.DefaultPadShiftKernelFFT;
+import net.imglib2.FinalDimensions;
+import net.imglib2.RandomAccessibleInterval;
+import net.imglib2.outofbounds.OutOfBoundsMirrorFactory;
+import net.imglib2.type.numeric.real.FloatType;
+import net.imglib2.view.Views;
+import org.jocl.NativePointerObject;
+import org.scijava.Context;
+import org.scijava.command.CommandService;
+import org.scijava.plugin.Plugin;
+
+import static net.haesleinhuepf.clijx.plugins.OpenCLFFTUtility.pad;
+
+@Plugin(type = CLIJMacroPlugin.class, name = "CLIJx_deconvolveRichardsonLucyFFT")
+public class DeconvolveRichardsonLucyFFT extends AbstractCLIJ2Plugin implements
+	CLIJMacroPlugin, CLIJOpenCLProcessor, OffersDocumentation, HasAuthor
+{
+
+	private static OpService ops;
+	private static Context ctx;
+	static {
+		// this initializes the SciJava platform.
+		// See https://forum.image.sc/t/compatibility-of-imagej-tensorflow-with-imagej1/41295/2
+		ctx = (Context) IJ.runPlugIn("org.scijava.Context", "");
+		if (ctx == null) ctx = new Context(CommandService.class, OpService.class);
+		ops = ctx.getService(OpService.class);
+	}
+
+	@Override
+	public boolean executeCL() {
+		boolean result = deconvolveRichardsonLucyFFT(getCLIJ2(), (ClearCLBuffer) (args[0]),
+			(ClearCLBuffer) (args[1]), (ClearCLBuffer) (args[2]), asInteger(args[3]));
+		return result;
+	}
+
+	public static boolean deconvolveRichardsonLucyFFT(CLIJ2 clij2, ClearCLBuffer input,
+													  ClearCLBuffer convolution_kernel, ClearCLBuffer destination, int num_iterations)
+	{
+
+		
+		ClearCLBuffer input_float = input;
+		if (input_float.getNativeType() != NativeTypeEnum.Float) {
+			input_float = clij2.create(input.getDimensions(), NativeTypeEnum.Float);
+			clij2.copy(input, input_float);
+		}
+
+		ClearCLBuffer convolution_kernel_float = convolution_kernel;
+		if (convolution_kernel.getNativeType() != NativeTypeEnum.Float) {
+			convolution_kernel_float = clij2.create(convolution_kernel
+					.getDimensions(), NativeTypeEnum.Float);
+			clij2.copy(convolution_kernel, convolution_kernel_float);
+		}
+
+		long start = System.currentTimeMillis();
+		
+		RandomAccessibleInterval imgF = clij2.pullRAI(input_float);
+
+		// compute extended dimensions based on image and PSF dimensions
+		long[] extendedSize = new long[imgF.numDimensions()];
+
+		for (int d = 0; d < imgF.numDimensions(); d++) {
+			extendedSize[d] = imgF.dimension(d) + convolution_kernel.getDimensions()[d];
+		}
+
+		FinalDimensions extendedDimensions = new FinalDimensions(extendedSize);
+
+		// extend image
+		RandomAccessibleInterval<FloatType> extended = (RandomAccessibleInterval) ops.run(DefaultPadInputFFT.class, imgF, extendedDimensions, false,
+						new OutOfBoundsMirrorFactory(OutOfBoundsMirrorFactory.Boundary.SINGLE));
```

Comment:
I took out the line where the PSF was extended with ops, because that is actually done in ```OpenCLUtility.pad```.    Should padding of the image also be done on GPU??   I timed it, and right now padding takes 400 ms on the bars test image, while 100 iterations of RL takes ~ 3000 ms.  So the padding is taking a significant  but not outrageous amount of time.  We could probably wait until the next iteration to implement padding on the GPU, however when we do we also need to make sure the final padding size is calculated based on the required FFT size (there is an op to calculate the next "fast" FFT size, which is automatically called by ```DefaultPadInputFFT``` which is critical for clFFT because it does not implement all sizes).

Reply:
Good that you ask. I attempted to do it on the GPU, but wasn't able to fully replicate the ops behaviour. The reason was: CLIJ uses EXTEND_BORDER behaviour and ops uses EXTEND_MIRROR. I was afraid that this is intentional and necessary. Furthermore, I wasn't 100% sure to which size the image should be extended. I traced the ops code down until there, but again, wasn't sure:
https://github.com/imagej/imagej-ops/blob/b7c48d01447ec26502a784b46b7ccb65bc246b90/src/main/java/net/imagej/ops/filter/fftSize/DefaultComputeFFTSize.java#L63-L66

Furthermore, does this also extend in Z? Is it necessary to extend in Z?